src/map/npc.cpp

```diff
@@ -4623,6 +4623,93 @@ int npc_script_event(struct map_session_data* sd, enum npce_event type){
 	return vector.size();
 }
 
+struct npc_data* dup_npc(npc_data* dnd, const char* name)
+{
+	npc_data* nd = npc_create_npc(dnd->bl.m, dnd->bl.x, dnd->bl.y);
+
+	nd->src_id = dnd->src_id ? dnd->src_id : dnd->bl.id;
+	nd->class_ = dnd->class_;
+	nd->speed = dnd->speed;
+	nd->bl.type = dnd->bl.type;
+	nd->subtype = dnd->subtype;
+
+	switch (nd->subtype) {
+
+	case NPCTYPE_SCRIPT:
+		++npc_script;
+		nd->u.scr.xs = dnd->u.scr.xs;
+		nd->u.scr.ys = dnd->u.scr.ys;
+		nd->u.scr.script = dnd->u.scr.script;
+		nd->u.scr.label_list = dnd->u.scr.label_list;
+		nd->u.scr.label_list_num = dnd->u.scr.label_list_num;
+		break;
+
+	case NPCTYPE_SHOP:
+	case NPCTYPE_CASHSHOP:
+	case NPCTYPE_ITEMSHOP:
+	case NPCTYPE_POINTSHOP:
+	case NPCTYPE_MARKETSHOP:
+		++npc_shop;
+		safestrncpy(nd->u.shop.pointshop_str, dnd->u.shop.pointshop_str, strlen(dnd->u.shop.pointshop_str));
+		nd->u.shop.itemshop_nameid = dnd->u.shop.itemshop_nameid;
+		nd->u.shop.shop_item = dnd->u.shop.shop_item;
+		nd->u.shop.count = dnd->u.shop.count;
+		nd->u.shop.discount = dnd->u.shop.discount;
+		break;
+
+	case NPCTYPE_WARP:
+		++npc_warp;
+		if (!battle_config.warp_point_debug)
+			nd->class_ = JT_WARPNPC;
+		else
+			nd->class_ = JT_GUILD_FLAG;
+		nd->u.warp.xs = dnd->u.warp.xs;
+		nd->u.warp.ys = dnd->u.warp.ys;
+		nd->u.warp.mapindex = dnd->u.warp.mapindex;
+		nd->u.warp.x = dnd->u.warp.x;
+		nd->u.warp.y = dnd->u.warp.y;
+		nd->trigger_on_hidden = dnd->trigger_on_hidden;
+		break;
+	}
+
+	if (nd->bl.m >= 0) {
+		map_addnpc(nd->bl.m, nd);
+		status_change_init(&nd->bl);
+		unit_dataset(&nd->bl);
+		nd->ud.dir = dnd->ud.dir;
+		npc_setcells(nd);
+		if (map_addblock(&nd->bl))
+			return nd;
+		if (nd->class_ != JT_FAKENPC) {
+			status_set_viewdata(&nd->bl, nd->class_);
+			if (map_getmapdata(nd->bl.m)->users)
+				clif_spawn(&nd->bl);
+		}
+	}
+	else {
+		map_addiddb(&nd->bl);
+	}
+	if (name == "") {
+		safesnprintf(nd->exname, ARRAYLENGTH(nd->exname), "dup_%d", nd->bl.id);
```

Comment:
A npc with the name autogenerated may already exist, please add a check

Reply:
done , and tested using 
```cs
prontera,155,153,0	script	inuse_test	445,{
	mes "inuse name test";
	if(select("start:close") == 2)
		end;
	.@original_npc$ = "my_test_npc2";
	.@new_npc$ = duplicate(.@original_npc$);
	.@id = getnpcid(0,.@new_npc$);
	.@new_npc$ = "dup_" + (.@id+2);
	duplicate(.@original_npc$,.@new_npc$);
	debugmes "Next Name: " + .@new_npc$;
	.@new_npc$ = duplicate(.@original_npc$);
	debugmes "New Name: " + .@new_npc$;
	//moving the npc to another location.
	getmapxy(.@m$,.@x,.@y,BL_NPC,.@original_npc$);
	if(!$@y) $@y = .@y;
	$@y -= 2;
	if(!(.@id = getnpcid(0,.@new_npc$))){
		mes "Something went wrong!";
		mes "The new npc is not found!";
		end;
	}
	unitwarp(.@id,.@m$,.@x,$@y);
	setnpcdisplay(.@new_npc$, "my test npc2", 445);
	mes "Done!";
end;
end;
}
prontera,155,149,0	script	test npc::my_test_npc2	444,{
	mes "My Name is: " + strnpcinfo(3);
	mes "OnInit event:";
	mes .text$;
end;
OnInit:
	.text$ = "it's me mario!!";
end;
}
```